### PR TITLE
EDM-649: Improve checks and error message for enrollment CSRs

### PR DESCRIPTION
**Pull request by lkatalin**
This PR adds a check that a CSR with a specific name does not already exist, and also clarifies the error message when an enrollment request cannot be auto-approved.

Additionally it removes two checks which seem extraneous (but let me know if we do need these):
- checking that the CN for an enrollment certificate is 16 chars long (AFAIU this name is allowed to be anything)
- ~~checking that the `Name` for a CSR resource is valid in kubernetes (AFAIU this name is allowed to be anything)~~

### CODE REVIEW

File: internal/service/certificatesigningrequest.go
Comment (hexfusion) [inline internal/service/certificatesigningrequest.go:245]: technically nil is also true here right?
```suggestion
	if err != nil {
	  if !errors.Is(flterrors.ErrResourceNotFound) {
```

File: api/v1alpha1/validation.go
Comment (fzdarsky) [inline api/v1alpha1/validation.go:239]: Please leave in. All of our API resources conform to this for consistency and compatibility with K8s.

**Review by fzdarsky — CHANGES_REQUESTED** [review]

File: internal/service/certificatesigningrequest.go
Comment (lkatalin) [reply internal/service/certificatesigningrequest.go:245]: In this case we want a nil error to evaluate to true because both a nil error and a non-ResourceNotFound error both mean the resource exists. ErrResourceNotFound is the only acceptable result of CertificateSigningRequest().Get(...) to execute the rest of the logic.

File: api/v1alpha1/validation.go
Comment (lkatalin) [reply api/v1alpha1/validation.go]: Okay no problem.

File: internal/service/certificatesigningrequest.go
Comment (hexfusion) [reply internal/service/certificatesigningrequest.go:245]: Ok then I would still recommend using errors.Is so that the error in the future could be `%w` wrapped and this eval would fail. Your call on that but it's a good practice in general.
```
if !errors.Is(err, flterrors.ErrResourceNotFound)
```

Comment (lkatalin) [reply internal/service/certificatesigningrequest.go]: Good suggestion, updated. I'm still getting used to this go syntax.

**Comment by lkatalin** [comment]
@hexfusion @fzdarsky Any other thoughts here? Is this good to merge after a rebase?

**Review by hexfusion — APPROVED** [review]

**Comment by lkatalin** [comment]
@fzdarsky Does this look fine to merge?

File: internal/service/certificatesigningrequest.go
Comment (fzdarsky) [inline internal/service/certificatesigningrequest.go:250]: A PUT verb on a resource has the behaviour "CreateOrUpdate", so it shouldn't be an error if the CSR already exists?

File: internal/service/certificatesigningrequest.go
Comment (fzdarsky) [inline internal/service/certificatesigningrequest.go:273]: A HTTP 400 (Bad Request) error is for when the API request isn't valid, not for when there are errors during handling of the request.

File: internal/service/certificatesigningrequest.go
Comment (lkatalin) [reply internal/service/certificatesigningrequest.go:250]: There are problems with auto-approving the CSR twice, so I will need to revisit this to see if there is another alternative besides throwing the error.

File: internal/service/certificatesigningrequest.go
Comment (lkatalin) [reply internal/service/certificatesigningrequest.go:273]: We could grow a 500 error here, but the API currently only has 200, 201, 400, 401, 404, and 409.

Comment (lkatalin) [reply internal/service/certificatesigningrequest.go:266]: @fzdarsky This is already a 400 error so that is not a new thing I am introducing in this PR. I am planning to submit a separate PR that updates the error codes where needed. Can this be merged as-is?

File: internal/service/certificatesigningrequest.go
Comment (lkatalin) [reply internal/service/certificatesigningrequest.go]: Update: I am going to make these changes in a separate PR.

**Comment by lkatalin** [comment]
Ping on merging this @fzdarsky @hexfusion :pray:

File: internal/service/certificatesigningrequest.go
Comment (fzdarsky) [reply internal/service/certificatesigningrequest.go:266]: Sure. The PR removes the length requirement, so that'll be helpful. It's mainly the change here that'll need to be undone in a follow-up.

**Review by fzdarsky — APPROVED** [review]

**Pull request merged**
